Add input port to the spring-mass system

amcastro-tri opened the pull request:
Learning how to setup input/output ports in System 2.0.
The plan is to create a Diagram with a PI controller on this plant. 

This is my little step towards implementing RBS (which essentially does F=m*a!).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2967)
<!-- Reviewable:end -->

amcastro-tri commented:
+@jwnimmer-tri for platform review.
+@david-german-tri and +@sherm1 for feature review.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNmQWmsC0tfep1fD4qm:394145968)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 214 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmQgsCekbyySUTSW0f:-KNmQgsCekbyySUTSW0g:1463018150) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L214)):*
> ```cpp
> }
> 
> TEST_F(SpringMassSystemTest, DyanmnicsWithExternalForce) {
> ```

Here I decided not to use Google's `TEST_P` since I din't want to test the fixture for all the above test with and without external input. Therefore I decided to remove rename `SetUp` to `Initialize` and call it explicitly. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Could you please rebase on latest master and resolve the merge conflicts with #2846?  I don't think the impact will be huge, but better to resolve it upfront.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNmSfyOkmqtpXSDpKXj:-1908376151)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint review -- everything except the unit test.

@amcastro-tri Please don't tag for platform review unless the cpplint build is passing (or you've run cpplint_wrapper.py by hand).  It only wastes my time pointing out nitpicks that the tool is designed to catch.

---

Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 89 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmUxrdU93Fg7Ptcy6t:-KNmUxreFlqKog4ZKg97:188484594) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L89)):*
> ```objc
>   /// mass.
>   SpringMassSystem(const std::string& name, double spring_constant_N_per_m,
>                    double mass_kg, bool system_is_forced = false);
> ```

Missing API doc (what does this new param even mean)?

---

*[drake/examples/spring_mass/spring_mass_system.h, line 113 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmW2fuOzg4s9z-JXBr:-KNmW2fuOzg4s9z-JXBs:160687164) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L113)):*
> ```objc
>   }
> 
>   double get_input_force(const MyContext& context) const {
> ```

Missing doxygen on public method.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 115 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmWKvkLChVAH4UTIit:-KNmWKvkLChVAH4UTIiu:86899719) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L115)):*
> ```objc
>   double get_input_force(const MyContext& context) const {
>     double external_force = 0;
>     if (context.get_num_input_ports() == 1) {
> ```

Prior to this line, consider `DRAKE_ASSERT(system_is_forced_ == (context.get_num_input_ports() == 1));` to cross-check the ctor setting with the `Context` passed in here.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 116 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVNfOx5FqR4Zum2Jy:-KNmVNfOx5FqR4Zum2Jz:1984637136) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L116)):*
> ```objc
>     double external_force = 0;
>     if (context.get_num_input_ports() == 1) {
>       const systems::VectorInterface<double> *input = context.get_vector_input(0);
> ```

Put the `*` on the typename; clang-format.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 116 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVS47z2sSXzqorSMG:-KNmVS47z2sSXzqorSMH:-2045096380) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L116)):*
> ```objc
>     double external_force = 0;
>     if (context.get_num_input_ports() == 1) {
>       const systems::VectorInterface<double> *input = context.get_vector_input(0);
> ```

Line exceeds 80 characters.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 121 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVqYpasWuUzjNLdok:-KNmVqYpasWuUzjNLdol:1766637010) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L121)):*
> ```objc
>       // So that the line below would simply read:
>       // external_force = context.get_input_vector(0).component(0)
>       // and becomes a one liner.
> ```

This comment block doesn't make sense.  I don't see how it relates to this implementation.  It it supposed to be a TODO?

---

*[drake/systems/framework/primitives/adder.cc, line 24 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmUJCSF6nq95Ej6o1I:-KNmUJCTnbzX_4D1XVHQ:-1497648900) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/systems/framework/primitives/adder.cc#L24)):*
> ```cpp
> 
>   System::add_port(OutputPort("Out", length));
> #endif
> ```

Remove commented-out code block.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNmVY86tOKSRA2FzdPf:2021105556)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass done.

---

Reviewed 1 of 4 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 83 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmYSZATaAnmSh9UmvN:-KNmYSZBElw5rY9YEHKg:666181885) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L83)):*
> ```cpp
>   // Used to test standalone systems not part of a Diagram.
>   static std::unique_ptr<FreestandingInputPort<double>> MakeInput(
>   std::unique_ptr<BasicVector<double>> data) {
> ```

Wrong indentation.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 85 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZIv46pUmAxpPIvM2:-KNmZIv46pUmAxpPIvM3:1628021212) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L85)):*
> ```cpp
>   std::unique_ptr<BasicVector<double>> data) {
>     return std::unique_ptr<FreestandingInputPort<double>>(
>         new FreestandingInputPort<double>(std::move(data)));
> ```

Consider `make_unique` for brevity.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 214 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmQgsCekbyySUTSW0f:-KNmYXCm0toTgZClCxEo:335980292) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L214)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Here I decided not to use Google's `TEST_P` since I din't want to test the fixture for all the above test with and without external input. Therefore I decided to remove rename `SetUp` to `Initialize` and call it explicitly. 
</details>
How about putting `setUp` back and having it call `Initialize(false);`, so all of the other code can stay undisturbed (and be less brittle).  Then just this one location can call `Initialize(true)` again.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZVEII0q5etoEgEGt:-KNmZVEJTEPFiAnOvVfZ:1119911543) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L217)):*
> ```cpp
>   // Initializes a spring mass system with an input port for an external force.
>   Initialize(true);
>   ASSERT_EQ(1, context_->get_num_input_ports());
> ```

Missing unit test coverage check somewhere that this value is zero when no external force (`Initialize(false)`) is used.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 239 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZunEEOF46Ff_FDcf:-KNmZunEEOF46Ff_FDcg:939526709) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L239)):*
> ```cpp
>               Eigen::NumTraits<double>::epsilon());
>   // The derivative of velocity is force over mass.
>   EXPECT_NEAR((-kSpring * 0.1 + kExternalForce )/ kMass,
> ```

Extra whitespace before `)`.
Missing whitespace before `/`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNm_-zxbDEZR4rNLt-H:-1777247566)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 4 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 89 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmUxrdU93Fg7Ptcy6t:-KNmytHRxQCcFsUX306_:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing API doc (what does this new param even mean)?
</details>
Done.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 113 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmW2fuOzg4s9z-JXBr:-KNmz9NuzGI9XIxZtZE0:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing doxygen on public method.
</details>
Done.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 115 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmWKvkLChVAH4UTIit:-KNmzemfUIzjd9X_UNlt:1357023505) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L115)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Prior to this line, consider `DRAKE_ASSERT(system_is_forced_ == (context.get_num_input_ports() == 1));` to cross-check the ctor setting with the `Context` passed in here.
</details>
Done. 

---

*[drake/examples/spring_mass/spring_mass_system.h, line 116 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVNfOx5FqR4Zum2Jy:-KNmzlN0WOKLwjSWOMuI:-116996864) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L116)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Put the `*` on the typename; clang-format.
</details>
Sorry, I forgot to push these before lunch and my meeting with Sherm. Next time ignore the PR if you see the red dot on CI's cpplint.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 116 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVS47z2sSXzqorSMG:-KNmzxwJ29px5YIZw7Aq:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L116)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Line exceeds 80 characters.
</details>
Done.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 121 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVqYpasWuUzjNLdok:-KNn-76pTCd189oIx_uG:317622596) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L121)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This comment block doesn't make sense.  I don't see how it relates to this implementation.  It it supposed to be a TODO?
</details>
Yes, it all one TODO. It says so at the top.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 83 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmYSZATaAnmSh9UmvN:-KNn-ZyzrXACVRfVKdnI:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L83)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Wrong indentation.
</details>
Done.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 85 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZIv46pUmAxpPIvM2:-KNn0Ptwm5nGx5hGcmwL:1756952154) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L85)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Consider `make_unique` for brevity.
</details>
Considered. Also in adder_test.cc

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 214 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmQgsCekbyySUTSW0f:-KNn0_GMiHqu48bUOGvB:-1319926881) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L214)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> How about putting `setUp` back and having it call `Initialize(false);`, so all of the other code can stay undisturbed (and be less brittle).  Then just this one location can call `Initialize(true)` again.
</details>
But in that case then `Initialize()` gets called twice for my test with an input port. Once within SetUp by google's unit testing framework and a second time explicitly by myself. 

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZVEII0q5etoEgEGt:-KNn1zgmskYg4jLfMv2F:1421350338) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L217)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test coverage check somewhere that this value is zero when no external force (`Initialize(false)`) is used.
</details>
Done. Added test on the very first unit test: `TEST_F(SpringMassSystemTest, Construction)`

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 239 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZunEEOF46Ff_FDcf:-KNn2N-sSAMFnP7vwPuQ:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L239)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Extra whitespace before `)`.
> Missing whitespace before `/`.
</details>
Done.

---

*[drake/systems/framework/primitives/adder.cc, line 24 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmUJCSF6nq95Ej6o1I:-KNn2PlCJDo28Qa8--Ft:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/systems/framework/primitives/adder.cc#L24)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Remove commented-out code block.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I addressed the reviewers' comments and merged with master. Thus this PR is ready for a second round of review.

---

Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNn2UkY1T4p7pAGCMp4:-715372916)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
This is nice!  I'm pleased to see it was such a compact change.

---

Review status: 0 of 4 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 124 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn0g6hHfa0VbageiUs:-KNn0g6hHfa0VbageiUt:-1729511111) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/spring_mass_system.h#L124)):*
> ```objc
>       // external_force = context.get_input_vector(0).component(0)
>       // and becomes a one liner.
>       external_force = input->get_value().x();
> ```

I had no idea `.x()` was a thing, but apparently it is, indeed, equivalent to `[0]`.  @sherm1 will be glad to hear I figured it out from Eigen's doxygen...eventually.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 214 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmQgsCekbyySUTSW0f:-KNn2wvyK8IuF7T3P6Ak:-1428096025) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L214)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> But in that case then `Initialize()` gets called twice for my test with an input port. Once within SetUp by google's unit testing framework and a second time explicitly by myself. 
</details>
Does that matter?

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 239 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZunEEOF46Ff_FDcf:-KNn2oqN3CLGz2DuJq6f:-1498054396) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L239)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done.
</details>
whitespace before `/` is still missing.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 26 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn1qGHioMRS8DuuqMs:-KNn1qGHioMRS8DuuqMt:-438169433) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/test/spring_mass_system_test.cc#L26)):*
> ```cpp
> 
> using std::unique_ptr;
> using std::make_unique;
> ```

nit: alphabetize

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 40 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn1s3ulq8UPSNwyS9i:-KNn1s3v0upJInPvLTQg:-438169433) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/test/spring_mass_system_test.cc#L40)):*
> ```cpp
> using systems::StateVector;
> using systems::FreestandingInputPort;
> using systems::SystemOutput;
> ```

nit: alphabetize

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 227 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn28iI8j8N5xltsfUn:-KNn28iI8j8N5xltsfUo:-172165302) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/test/spring_mass_system_test.cc#L227)):*
> ```cpp
> 
>   // Creates a free standing input port not actually connected to the output of
>   // another system (typical case) but that has its own data in force_vector.
> ```

BTW, I get what you mean, but the parenthetical here is kind of confusing.  Consider dropping those two words - I think it's clear enough without them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNn31ni9wOfDYW5RDVX:506959429)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 89 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmUxrdU93Fg7Ptcy6t:-KNn63720f6C3YW_adZQ:-719066722) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L89)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done.
</details>
Not done in r3.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 113 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmW2fuOzg4s9z-JXBr:-KNn64wEYzq9UdiplXUI:-719066722) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L113)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done.
</details>
Not done in r3.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 121 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmVqYpasWuUzjNLdok:-KNn6ExtzjI0SiFxNOlK:-335939034) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/spring_mass_system.h#L121)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Yes, it all one TODO. It says so at the top.
</details>
Ah.  When I first read it, the TODO was not present / pushed yet.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 83 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmYSZATaAnmSh9UmvN:-KNn3GKkD53uzAIRTPa0:1437640912) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L83)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done.
</details>
In r3, the marked line is still unindented; it should be indented by at least two spaces (and probably 4 or more).

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 214 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmQgsCekbyySUTSW0f:-KNn6cuFBxpkoQtxEgJT:728519757) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L214)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Does that matter?
</details>
I (as well) do not see why doing it twice would be a problem?

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 217 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZVEII0q5etoEgEGt:-KNn5itTzYo6HR3PAfXK:2124395609) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L217)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done. Added test on the very first unit test: `TEST_F(SpringMassSystemTest, Construction)`
</details>
I am not seeing this in r3?

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 239 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmZunEEOF46Ff_FDcf:-KNn5nH0FFGl45ZbcN1I:437475384) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L239)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> whitespace before `/` is still missing.
</details>
Still missing in r3.  Given how many thins are marked done but are not, perhaps a `git push` didn't happen yet?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I am sorry, I forgot to push the latest changes. PTAL.

jwnimmer-tri commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 105 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnC03gv3d9h88l4H98:-KNnC03gv3d9h88l4H99:-1438276373) ([raw file](https://github.com/robotlocomotion/drake/blob/0343bf3f3f3503213b144cdc092b873de0ab6ac2/drake/examples/spring_mass/test/spring_mass_system_test.cc#L105)):*
> ```cpp
>   Initialize();
>   // Asserts zero inputs for this case.
>   ASSERT_EQ(0, context_->get_num_input_ports());
> ```

Unclear why this has to be `ASSERT` instead of `EXPECT`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 214 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNmQgsCekbyySUTSW0f:-KNnFeNnTa7j7u01CkeP:-1622819128) ([raw file](https://github.com/robotlocomotion/drake/blob/06081acbd085a0bd4451a6f7f63fdd5cd3081bed/drake/examples/spring_mass/test/spring_mass_system_test.cc#L214)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I (as well) do not see why doing it twice would be a problem?
</details>
Good point. Done.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 26 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn1qGHioMRS8DuuqMs:-KNnDgWgSjUFzCzWm3Ge:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/test/spring_mass_system_test.cc#L26)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> nit: alphabetize
</details>
Done.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 40 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn1s3ulq8UPSNwyS9i:-KNnDy-PqqTHni7LF5NO:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/test/spring_mass_system_test.cc#L40)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> nit: alphabetize
</details>
Done.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 105 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnC03gv3d9h88l4H98:-KNnE4sWLj2LCH1k0grM:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/0343bf3f3f3503213b144cdc092b873de0ab6ac2/drake/examples/spring_mass/test/spring_mass_system_test.cc#L105)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unclear why this has to be `ASSERT` instead of `EXPECT`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 268 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGJiVGSAEsvIL_j59:-KNnGJiVGSAEsvIL_j5A:-429309234) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/spring_mass_system.h#L268)):*
> ```objc
>   const double spring_constant_N_per_m_;
>   const double mass_kg_;
>   const bool system_is_forced_{false};
> ```

Consider whether this is necessary

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNnGYzsi1IXbpNY4X6m:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 268 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGJiVGSAEsvIL_j59:-KNnGvLUSAmcciczziVa:406695218) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/spring_mass_system.h#L268)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Consider whether this is necessary
</details>
The `{}` are required by the drake style guide; the `false` is a bit weird, I agree.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 56 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGjbKICiSnRg3Vpym:-KNnGjbKICiSnRg3Vpyn:1810493545) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/test/spring_mass_system_test.cc#L56)):*
> ```cpp
>   }
> 
>   void Initialize(bool with_input_force = false) {
> ```

Nobody is using this default argument anymore.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNnH47tQ0jkWMW7jJDk:1417088721)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 268 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGJiVGSAEsvIL_j59:-KNnGJiVGSAEsvIL_j5A:564900562) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/spring_mass_system.h#L268)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The `{}` are required by the drake style guide; the `false` is a bit weird, I agree.
</details>
I didn't actually mean to publish this.  Please disregard!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
> Consider whether this is necessary

I didn't mean to publish this at all - please disregard.  (I would have made this comment on Reviewable.io but it's having some kind of outage.)

amcastro-tri commented:
@david-german-tri: Form the very humble perspective of a user, let me give you some feedback on my feeling implementing this simple input port to the mass-spring system. 
I found the abstractions you designed in System 2.0 to be quite beautiful and very useful to split important concepts.
I must say however that the API for the user (like me) is not quite there yet. It took me two full days of study and a f2f conversation with you to get this started. And even still I find myself uneasy in a number of points:
- Doing quite a bit of unique_ptr shuffling in a number of places. I think we should not expect our user to do this.
- I find it strange that in my test `DyanmnicsWithExternalForce` I have to add an input port to the context. In principle the SpringMassSystem should already have this information encoded. The context should only be used for actual data (ports and systems only hold the constant connectivity). This goes along #2955.
- Notice that for this implementation to add an input port I need to do it in two places: AllocateDefaultContext and in the test itself.
- I feel quite uneasy writing lengthy things like `context.get_vector_input(0)->get_value().x()` to just get a scalar value out of the context. I don't think our users will like it either. 

Again, I think the System abstraction are just great. However, I do think there is work to do with the concepts of ports. It is not only a matter of API but of splitting the "connectivity" concept form the actual data flowing through those connections (the user should not need to add ports to the context but to the system). 

These comments might seem subjective so I will try to quantify them. When reading System 3.0 I had a pretty good idea how to set a system up and connects things in less than an hour. It took me two days and a f2f conversation with @david-german-tri to get this started. 

I will try to come up with a simulink example of a simple PID controlled mass-spring system and how this setup could look in our System 2.0, hopefully that will help to clarify my observations.

I hope this is somehow useful!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 268 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGJiVGSAEsvIL_j59:-KNnHT_KFbLhrU0GDsEN:765371266) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/spring_mass_system.h#L268)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The `{}` are required by the drake style guide; the `false` is a bit weird, I agree.
</details>
I don't understand the comment here. You mean because it is already initialized in the constructor?
The `false` to me is nicer because clearly shows the default. I matter of style I guess.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 56 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGjbKICiSnRg3Vpym:-KNnHq1TzMLnU_8CJVli:-1642843890) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/test/spring_mass_system_test.cc#L56)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Nobody is using this default argument anymore.
</details>
Now yes :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNnIBOzBNUJrdhanViG:453670549)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/examples/spring_mass/spring_mass_system.h, line 268 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnGJiVGSAEsvIL_j59:-KNnRr7c4nRA2aGXpNCu:202782889) ([raw file](https://github.com/robotlocomotion/drake/blob/569e504898f48eceff72bcc9f2b46aac273ac61d/drake/examples/spring_mass/spring_mass_system.h#L268)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I don't understand the comment here. You mean because it is already initialized in the constructor?
> The `false` to me is nicer because clearly shows the default. I matter of style I guess.
> 
</details>
I say that the `false` is weird because is has zero effect -- in every case, the constructor sets some other value.  It could say `peanut` and be just as descriptive.  Saying the default in the ctor's default argument, and in the member field, is brittle and misleading.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed -- some review comments and some editorial comments about the API.

---

Reviewed 2 of 5 files at r3, 2 of 3 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 124 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn0g6hHfa0VbageiUs:-KNnylzrq6mo3SZ7c6U5:-883441687) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/spring_mass_system.h#L124)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I had no idea `.x()` was a thing, but apparently it is, indeed, equivalent to `[0]`.  @sherm1 will be glad to hear I figured it out from Eigen's doxygen...eventually.
</details>
+1 for doxygen!

---

*[drake/examples/spring_mass/spring_mass_system.h, line 131 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnz68g2jNF2UJrkHWi:-KNnz68g2jNF2UJrkHWj:-880163532) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/spring_mass_system.h#L131)):*
> ```objc
>       // So that the line below would simply read:
>       // external_force = context.get_input_vector(0).component(0)
>       // and becomes a one liner.
> ```

BTW, I expect this will get cleaner when System 2 support abstract value types for ports. Then we'll need support for downcasting those to their actual types anyway so we could have something like `output_port.get_value_as<T>(context)`.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 267 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnzzj8Sa6mMVxrpbus:-KNnzzj8Sa6mMVxrpbut:1655181394) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/spring_mass_system.h#L267)):*
> ```objc
>   const std::string name_;
>   const double spring_constant_N_per_m_;
>   const double mass_kg_;
> ```

These doubles should have initializers.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 86 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNo-SLDjwiNGvp5hvXS:-KNo-SLDjwiNGvp5hvXT:-1019452237) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/test/spring_mass_system_test.cc#L86)):*
> ```cpp
>   // not connected to any other output port in the system.
>   // Used to test standalone systems not part of a Diagram.
>   static std::unique_ptr<FreestandingInputPort<double>> MakeInput(
> ```

BTW, hopefully these can be replaced by a "Constant" system block with one output port that we could simply plug into an input port to give it a value.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 160 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNo-rllyAe69gOOGzGU:-KNo-rllyAe69gOOGzGV:209947105) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/test/spring_mass_system_test.cc#L160)):*
> ```cpp
>   // To minimize user's typing.
>   ContinuousState<double>* continuous_state =
>       context_->get_mutable_state()->continuous_state.get();
> ```

BTW, better I think if we can get rid of these extra classes altogether. A continuous state is inherently a vector of scalars -- if we can manage to allocate them contiguously in a diagram context then we can simply return Eigen vectors.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 211 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNo094QHNcD0QokI9Nj:-KNo094RS-c9udhWvnnp:678074825) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/test/spring_mass_system_test.cc#L211)):*
> ```cpp
> }
> 
> TEST_F(SpringMassSystemTest, DyanmnicsWithExternalForce) {
> ```

typo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNo0VWpmM5_UfbD2bI9:1972722458)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Addressed last set of reviews. PTAL.

---

Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/spring_mass/spring_mass_system.h, line 131 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnz68g2jNF2UJrkHWi:-KNqkGXy5X4rYMvVnywQ:1655918653) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/spring_mass_system.h#L131)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, I expect this will get cleaner when System 2 support abstract value types for ports. Then we'll need support for downcasting those to their actual types anyway so we could have something like `output_port.get_value_as<T>(context)`.
</details>
Yes, I agree that'd look beautiful. 

---

*[drake/examples/spring_mass/spring_mass_system.h, line 267 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNnzzj8Sa6mMVxrpbus:-KNql4l_Jgfzif9AGGDm:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/spring_mass_system.h#L267)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> These doubles should have initializers.
</details>
Done.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 227 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNn28iI8j8N5xltsfUn:-KNqmVtAl2H2IFVE3cCO:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/70d1de8838b7497106af32b73c0e97f00e7b1720/drake/examples/spring_mass/test/spring_mass_system_test.cc#L227)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, I get what you mean, but the parenthetical here is kind of confusing.  Consider dropping those two words - I think it's clear enough without them.
</details>
Done.

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 86 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNo-SLDjwiNGvp5hvXS:-KNql9NqnXFmbjv43GZw:-140206566) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/test/spring_mass_system_test.cc#L86)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, hopefully these can be replaced by a "Constant" system block with one output port that we could simply plug into an input port to give it a value.
</details>
I agree on this also. This test then would reduce to having a system with two sub-systems: a `Constant` system and the `SpringMassSystem` with the output of `Constant` connected to the input of `SpringMassSystem`

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 160 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNo-rllyAe69gOOGzGU:-KNqlazRcTiEhK_n69qG:1771366754) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/test/spring_mass_system_test.cc#L160)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, better I think if we can get rid of these extra classes altogether. A continuous state is inherently a vector of scalars -- if we can manage to allocate them contiguously in a diagram context then we can simply return Eigen vectors.
</details>
+1 to simply returning Eigen vectors. @david-german-tri, what do you think?

---

*[drake/examples/spring_mass/test/spring_mass_system_test.cc, line 211 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-KNo094QHNcD0QokI9Nj:-KNqm7aXhYqdRREll2LP:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/c49fec2e26cb1ad722a7e09d988187236e7898c9/drake/examples/spring_mass/test/spring_mass_system_test.cc#L211)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> typo
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967#-:-KNqmbLebmAu1pAez5YI:486455397)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 2 files at r7.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2967)*
<!-- Sent from Reviewable.io -->
